stl/inc/xutility

```diff
@@ -4823,24 +4852,130 @@ public:
+    }
+
+#endif // ^^^ intrinsics unavailable ^^^
+
+    // using multiply-shift instead of modulo to fairly map values in range [0, _Mod), adapted from arXiv:1805.10941v4
+    static constexpr _Udiff _Reduce(_Udiff _Value, _Udiff _Mod, size_t _Bitrange) noexcept {
+        constexpr auto _Udiff_bits = CHAR_BIT * sizeof(_Udiff);
+        constexpr auto _Uint_bits  = CHAR_BIT * sizeof(uint32_t);
+        // _Bitrange may not exceed the number of bits in result type
+        _STL_INTERNAL_CHECK(_Bitrange <= _Udiff_bits);
+        if (_Bitrange > _Uint_bits) {
+            uint64_t _Val_hi = 0;
+            uint64_t _Val_lo = __rng_umul128(_Value, _Mod, &_Val_hi);
+
+            if (_Bitrange < _Udiff_bits) {
+                _Val_hi = __rng_shiftright128(_Val_lo, _Val_hi, static_cast<uint32_t>(_Bitrange));
             }
+            return static_cast<_Udiff>(_Val_hi);
+        } else if (_Bitrange > _Uint_bits / 2) {
+            // ensure promotion to ull to keep all bits
+            return static_cast<_Udiff>((static_cast<uint64_t>(_Value) * _Mod) >> _Bitrange);
+        }
+
+        // avoid unnecessary promotion to 64-bit for smaller types
+        return static_cast<_Udiff>((static_cast<uint32_t>(_Value) * _Mod) >> _Bitrange);
+    }
+
+    _NODISCARD static constexpr bool _Is_power_of2(_Udiff _Value) noexcept {
+        return _Value && !(_Value & (_Value - 1));
+    }
 
-            // _Ret is [0, _Mask], _Index - 1 <= _Mask, return if unbiased
-            if (_Ret / _Index < _Mask / _Index || _Mask % _Index == _Udiff(_Index - 1)) {
-                return static_cast<_Diff>(_Ret % _Index);
+    _NODISCARD _Diff operator()(const _Diff _Index) { // adapt _Urng closed range to [0, _Index)
+        constexpr auto _Udiff_bits = CHAR_BIT * sizeof(_Udiff);
+        const auto _UIndex         = static_cast<_Udiff>(_Index);
+        _Udiff _Ret                = _Get_bits(); // random bits
+        _Udiff _Mask               = _Bmask; // 2^N - 1, _Ret is within [0, _Mask]
+        size_t _Bitrange           = _Bits;
+
+        while (_Mask < static_cast<_Udiff>(_Index - 1)) { // need more random bits
+            // if loop is entered, _Bits can't shift out of range
+            _Ret <<= _Bits;
+            _Ret |= _Get_bits();
+            _Mask <<= _Bits;
+            _Mask |= _Bmask;
+            _Bitrange += _Bits;
+        }
+
+        // exceeding bits from _Get_bits() that don't fit into _Ret are discarded,
+        // therefore _Bitrange can never be larger than _Udiff_bits
+        if (_Bitrange > _Udiff_bits) {
+            _Bitrange = _Udiff_bits;
+        }
+
+        _Udiff _Threshold = _Mask - (_Mask % _UIndex);
+        for (;;) { // try a sample random value
+            // _Ret is [0, _Mask], _UIndex - 1 <= _Mask, return if unbiased
+            if (_Ret < _Threshold || _Is_power_of2(_UIndex)) {
+                return static_cast<_Diff>(_Reduce(_Ret, _UIndex, _Bitrange));

```

Comment:
This rejection criteria is incorrect for multiply-shift.

Assume that we'd like to generate a uniform random integer in [0, _OutRange_) with a uniform random integer generator in [0, _InRange_), and that _OutRange_ <= _InRange_.

Let _MapQuo_ = intdiv(_InRange_, _OutRange_), _MapRem_ = mod(_InRange_, _OutRange_). Ideally, we'd like to map _MapQuo_ input values to each output value to make the output uniform. A total of _MapQuo_ * _OutRange_ = _InRange_ - _MapRem_ input values are accepted, and the remaining _MapRem_ are rejected.

Let _X_ be an integer in [0, _InRange_), _Y_ = _X_ * _OutRange_, _YQuo_ = intdiv(_Y_, _InRange_), _YRem_ = mod(_Y_, _InRange_). Thus _YQuo_ ∈ [0, _OutRange_), _YRem_ ∈ [0, _InRange_).

If we accept _YRem_ >= _MapRem_ and reject _YRem_ < _MapRem_, there will be exactly _MapQuo_ input values _X'_ producing _YQuo'_ = _YQuo_, _YRem'_ ∈ [_MapRem_, _InRange_) and _YRem'_ ≡ _YRem_ (mod _OutRange_).

Possible implementation:
```
const auto _Out_max = static_cast<_Udiff>(_Index - 1);

// multiply-shift
_Udiff _Hi = /* (_Ret * _UIndex) >> _Bitrange */;
_Udiff _Lo = /* (_Ret * _UIndex) & _Mask */;

// save a division when _Lo >= _Index - 1 >= _Threshold
if (_Lo < _Out_max) {
    return static_cast<_Diff>(_Hi);
}

// rejection threshold
const auto _Threshold = static_cast<_Udiff>((_Mask - _Out_max) % _UIndex);
for (;;) {
    if (_Lo >= _Threshold) {
        return static_cast<_Diff>(_Hi);
    }
    _Ret = /* get _Bitrange random bits */;
    _Hi = /* (_Ret * _UIndex) >> _Bitrange */;
    _Lo = /* (_Ret * _UIndex) & _Mask */;
}
```

Reply:
You're right. Thanks for pointing it out and the additional explanation. I will change it accordingly.